### PR TITLE
8282045: When loop strip mining fails, safepoints are removed from loop anyway

### PR DESCRIPTION
**Pull request by rwestrel**
I noticed that if loop strip mining fails because a safepoint is not
found right above the exit test (following partial peel for instance),
all safepoints are removed from the loop anyway. That's fixed by the
change in IdealLoopTree::counted_loop() where rather than test if loop
strip mining is enabled, the check now verifies that loop strip mining
was successful.

With that change,
compiler/c2/irTests/TestAutoVectorization2DArray.java fails. The loop
is not converted into a strip mined loop because there's no safepoint
above the exit test after partial peeling. The loop strip mining logic
is too strict when it comes to the safepoint location. Any safepoint
that dominates the exit and is in the loop as long as there's no side
effect between the safepoint and the exit can be used. The patch
implements that change as well. TestAutoVectorization2DArray.java
passes as a result.

The existing requirement to have no safepoint on the backedge is too
strict as well. If the loop has another safepoint that can be used for
strip mining, then the safepoint on the backedge can safely be
dropped. That's also implemented by the patch.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8282045](https://bugs.openjdk.java.net/browse/JDK-8282045): When loop strip mining fails, safepoints are removed from loop anyway

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Christian Hagedorn](https://openjdk.java.net/census#chagedorn) (@chhagedorn - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7513/head:pull/7513` \
`$ git checkout pull/7513`

Update a local copy of the PR: \
`$ git checkout pull/7513` \
`$ git pull https://git.openjdk.java.net/jdk pull/7513/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7513`

View PR using the GUI difftool: \
`$ git pr show -t 7513`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7513.diff">https://git.openjdk.java.net/jdk/pull/7513.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/opto/loopnode.cpp [hunk 1]
```diff
@@ -1859,6 +1849,38 @@ bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_
     }
   }
 
+  Node *sfpt2 = NULL;
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopnode.cpp:1852]:
```suggestion
  Node* sfpt2 = NULL;
```

File: src/hotspot/share/opto/loopnode.cpp [hunk 2]
```diff
@@ -1859,6 +1849,38 @@ bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_
...
+    sfpt2 = find_safepoint(back_control, x, loop);
+  } else {
+    sfpt2 = iff->in(0);
+    if (sfpt2->Opcode() != Op_SafePoint) {
+      sfpt2 = NULL;
+    }
+  }
+
+  if (x->in(LoopNode::LoopBackControl)->Opcode() == Op_SafePoint) {
+    Node *sfpt = x->in(LoopNode::LoopBackControl);
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopnode.cpp:1863]:
```suggestion
    Node* sfpt = x->in(LoopNode::LoopBackControl);
```

File: src/hotspot/share/opto/loopnode.cpp (hunk 1 above)
Comment (TobiHartmann) [reply src/hotspot/share/opto/loopnode.cpp:1852]: I would suggest to rename `sfpt2` to `sfpt` and `sfpt` below to `backedge_sfpt`.

File: src/hotspot/share/opto/loopnode.cpp [hunk 3]
```diff
@@ -1859,6 +1849,38 @@ bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_
...
+    if (((iv_bt == T_INT && LoopStripMiningIter != 0) ||
+         iv_bt == T_LONG) &&
+        sfpt2 == NULL) {
+      // Leaving the safepoint on the backedge and creating a
+      // CountedLoop will confuse optimizations. We can't move the
+      // safepoint around because its jvm state wouldn't match a new
+      // location. Give up on that loop.
+      return false;
+    }
+    if (deleteable) {
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopnode.cpp:1874]: Can't you check `is_deleteable_safept` inline here like old code did?

**Review by TobiHartmann — COMMENTED** [review]
Noticed a typo in `PhaseIdealLoop::find_safepoint` that you might want to fix as well with this patch:
`there's not side effect` -> `there's no side effect

**Comment by rwestrel** [comment]
@TobiHartmann  thanks for looking at this. The new commit should take care of your comments.

**Review by TobiHartmann — APPROVED** [review]
Thanks, looks good to me.

**Comment by TobiHartmann** [comment]
Also, I executed some testing and it all passed.

**Comment by rwestrel** [comment]
@TobiHartmann @chhagedorn thanks for the reviews.

**Pull request closed without merging**